### ISSUE TITLE
Syncing users episode progress locally (getting episode ID's)

### ISSUE THREAD

**Issue by nixgates** [trakt/api-help#75]
Afternoon, 

Just a quick question to see if it's a possibility to add the episode ID's to the sync/watched/shows endpoint or is there a better way to obtain the ID's of the episodes a user has watched than I'm currently doing?

Currently my app's database is stored with the trakt ID of each object as the unique field. Issue being that if I was to do an initial fresh sync of a users watched progress, I need to request the above endpoint and then identify each episodes ID to store them down.

The issue is possible on my end in the method I'm storing these items where I could possibly use a mutation of the show ID to store episodes and seasons, but I would love to be able to acquire the episode ID for each episode in one hit and make a larger single database transaction than multiple smaller ones for the obvious speed benefits.

**Comment by rectifyer** [comment]
We don't have a method that provides this info yet, but it is something I am working on. I want to add super minimal methods that only return the Trakt `id` along with the `watched_at` timestamps and that is it. I'm still working on the JSON structure and will post my proposal here once I have some examples.

**Comment by rectifyer** [comment]
Quick followup about what the minimal amount of data is needed to make this usable for your app. I think it only needs the `id` and the `watched_at` timestamps right?

**Comment by nixgates** [comment]
@rectifyer I did read somewhere before about you wanting to do this and I think it's a brilliant idea as it will likely reduce some strain on your servers as well.

For most general endpoints "Lists, Trending, Popular" I would think a simple list of ID's would suit it perfectly. My only concern is for the shows sync endpoint it may be pertinent to include the season/episode numbers. Without that information, the database entries would not be particularly useful.

In example, an application would be unable to build a list of the episodes a user is up to without processing the meta for all of the shows episode entries which kind of defeats the purpose in my eyes.

So in short, for the shows sync endpoint I'd prefer to see a format similar to (tried to keep it as minimal as possible):

```
[
    {
        "1234" :             #Trakt Show ID
            "episodes: [
                {
                "1234":{          # Episode ID
                    "season": 1, 
                    "episode": 1
                }
            ]
        }
    }
]
```

**Comment by rectifyer** [comment]
I wasn’t think of including the season + episode numbers and just sticking to IDs only. If you are caching the the entire show, wouldn’t you already have all the episode IDs cache locally in the app?

This is a work in progress, but here is what I came up with (prior to your response). The keys would be the Trakt ID and the array is all the time stamps when the item was watched. This actually contains a bit more info since you can get the exact times it was watched. With this info, you can key all the watched items (the keys), the play count (size of the array), last watched (newest time stamp), and the full watched history for each item. Having the time stamp is important since apps might want to indicate when they last watched the item, similar to how the Trakt website does it.

For movies and episodes (no shows):
`{
  "1": [
    "2014-10-11T17:00:54.000Z",
    "2015-10-11T17:00:54.000Z",
    "2016-10-11T17:00:54.000Z"
  ],
  "2": [
    "2014-10-11T17:00:54.000Z",
    "2015-10-11T17:00:54.000Z",
    "2016-10-11T17:00:54.000Z"
  ]
}`

For shows, it would have a top level key which is the show ID.
`{
  "1": {
    "11": [
      "2014-10-11T17:00:54.000Z",
      "2015-10-11T17:00:54.000Z",
      "2016-10-11T17:00:54.000Z"
    ],
    "12": [
      "2014-10-11T17:00:54.000Z",
      "2015-10-11T17:00:54.000Z",
      "2016-10-11T17:00:54.000Z"
    ]
  }
}`

I am also not sure if this should be an entirely new method or the same method name with `?extended=min` as a parameter. It might be confusing to use the same method name since the output is so different?[](url)

**Comment by nixgates** [comment]
I think on the point of setting it as a parameter or not, adding a min option to the extended parameter is fine, as long as it is documented well inside the API docs and I personally wouldn't think any different of it.

Funnily enough, while writing an explanation on my applications database I've actually identified a much faster method of performing an initial sync with your servers than I'm currently employing so I believe I've ended up solving my own silly mistake. 

In regards to moving forward with the minimal watched endpoints I think you are on the money though.

I'll close this issue off and try and look at my problems a little more closely in the future 😂

**Comment by rectifyer** [comment]
@nixgates nice, I'm curious what you're doing to optimize?

**Comment by nixgates** [comment]
My application has three base functions for handling record insertion into the respective tables for items that are unseen.

The initial sync used these functions and it would simply insert the minimum info during a sync eg: Episode (episode_id, show_id, season, episode) per record when performing the initial sync. Unfortunately because the trakt_id wasn't being supplied to the episode function it needed to poll the trakt servers for that information.  This means the whole process was slowed through network requests and individual database transactions.

What I've decided to do now is process any syncs independent of these functions. So now it pulls the watched/collected episodes, then polls '/shows/?/seasons?extended=episodes' endpoint for the complete episode list for each show required. Then it compares the two sets of data per show and builds a single database transaction.
